### PR TITLE
SAK-42905 Gradebook: Add comments to course grade

### PR DESCRIPTION
**Pull request by jkozar2**
This PR adds the ability to comment on Course Grades in Gradebook, in a way that looks and works just like regular assignment comments. They are stored and retrieved the same way as well, only altering the existing workflow to allow for them to be attached to the Course Grade rather than a normal item.
https://sakaiproject.atlassian.net/browse/SAK-42905

### CODE REVIEW

File: gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java [hunk 1]
```diff
@@ -2759,6 +2759,15 @@ private void postUpdateGradeEvent(String gradebookUid, String assignmentName, St
                 "/gradebook/" + gradebookUid + "/" + assignmentName + "/" + studentUid + "/" + pointsEarned + "/student");
     }
 
+    /**
+     * Get the student's course grade's GradableObject ID.
+     *
+     * @return coursegrade's GradableObject ID.
+     */
+    public Long getCourseGradeId(Long gradebookId){
```
Comment (adrianfish) [inline gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java:2767]:
```suggestion
    public Long getCourseGradeId(Long gradebookId) {
```

File: gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java [hunk 2]
```diff
@@ -4775,18 +4784,24 @@ public CommentDefinition getAssignmentScoreComment(String gradebookUid, Long ass
         if (StringUtils.isBlank(gradebookUid) || assignmentId == null || StringUtils.isBlank(studentUid)) {
             throw new IllegalArgumentException("gradebookUid, assignmentId and studentUid must be valid.");
         }
-
+        String assignmentName = "";
         final GradebookAssignment assignment = getAssignmentWithoutStats(gradebookUid, assignmentId);
         if (assignment == null) {
-            throw new AssessmentNotFoundException("There is no assignmentId " + assignmentId + " for gradebookUid " + gradebookUid);
+            CourseGrade courseGrade = getCourseGrade(getGradebook(gradebookUid).getId());
+            if(courseGrade != null && courseGrade.getId().equals(assignmentId)){   //check if this is a course grade before declaring it Not Found
```
Comment (adrianfish) [inline gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java:4791]: I'd like to see a unit test to cover the changes in this PR. There's a lot more testing in the grading service after my big refactor, so it's easier to add them

**Review by adrianfish — DISMISSED** [review]
We need some tests on this stuff, the testing for a coursegrade bits especially.

File: gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java (hunk 2 above)
Comment (jkozar2) [reply gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java:4791]: I added one [GradingServiceTests.courseGradeComment] closely following the example of your assignmentScoreComment test. Please let me know of any others you'd like added!

**Comment by ern** [comment]
Moving this to a draft until conflicts are resolved

**Comment by jkozar2** [comment]
Just merged in Sakaiproject/master to this SAK-42905 feature branch to resolve the merge conflicts.

**Comment by jkozar2** [comment]
Noticed the comment bubble wouldn't appear during testing, just fixed that up.

**Review by ern — APPROVED** [review]

**Comment by ern** [comment]
Thanks Joe for getting this updated

**Pull request merged**